modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java

```diff
@@ -131,4 +139,99 @@ public void process(ResultSet<SqlRow> resultSet) {
             resultSet.forEachRemaining(res::add);
         }
     }
+
+    @Test
+    public void testEarlyQueryTimeout() {
+        Statement stmt = igniteSql().statementBuilder()
+                .query("SELECT * FROM TABLE(SYSTEM_RANGE(1, 1000000000000000))")
+                .queryTimeout(1, TimeUnit.MILLISECONDS)
+                .build();
+
+        // Do not have enough time to do anything.
+        assertThrowsSqlException(Sql.EXECUTION_CANCELLED_ERR, "Query timeout", () -> {
+            igniteSql().execute(null, stmt);
+        });
+    }
+
+    @Test
+    public void testQueryTimeout() {
+        Statement stmt = igniteSql().statementBuilder()
+                .query("SELECT * FROM TABLE(SYSTEM_RANGE(1, 1000000000000000))")
+                .queryTimeout(100, TimeUnit.MILLISECONDS)
+                .build();
+
+        ResultSet<SqlRow> rs;
+        while (true) {
+            try {
+                rs = igniteSql().execute(null, stmt);
+                break;
+            } catch (SqlException e) {
+                if (e.code() == Sql.PLANNING_TIMEOUT_ERR || e.code() == Sql.EXECUTION_CANCELLED_ERR) {
+                    continue;
+                }
+                fail(e.getMessage());
+            }
+        }
+        ResultSet<SqlRow> resultSet = rs;
+        assertNotNull(resultSet);
+
+        // Read data until exception occurs.
+        assertThrowsSqlException(Sql.EXECUTION_CANCELLED_ERR, "Query timeout", () -> {
+            while (resultSet.hasNext()) {
+                resultSet.next();
+            }
+        });
+    }
+
+    @Test
+    public void testQueryTimeoutIsPropagatedFromTheServer() throws Exception {
```

Comment:
I see that this test sometimes produces NPE during cancellation
```
[2024-06-24T10:38:03,199][WARN ][%issat_n_0%sql-planning-pool-1][task] Volcano planning times out, cancels the subsequent optimization.
[2024-06-24T10:38:03,446][ERROR][%issat_n_0%sql-execution-pool-1][FailureProcessor] Critical system error detected. Will be handled accordingly to configured handler [hnd=NoOpFailureHandler [], failureCtx=FailureContext [type=CRITICAL_ERROR, err=org.apache.ignite.lang.IgniteException: IGN-CMN-65535 TraceId:05602438-1bba-4f45-94e9-e63ab8c6d21f Unexpected error during execute fragment 145 of query 6fe450b1-b2fb-45a8-9774-c1af41e7e82c]]
 org.apache.ignite.lang.IgniteException: Unexpected error during execute fragment 145 of query 6fe450b1-b2fb-45a8-9774-c1af41e7e82c
	at org.apache.ignite.internal.sql.engine.exec.QueryTaskExecutorImpl.lambda$execute$0(QueryTaskExecutorImpl.java:96) ~[main/:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.NullPointerException
	at org.apache.ignite.internal.sql.engine.exec.ExecutionServiceImpl$DistributedQueryManager.acknowledgeFragment(ExecutionServiceImpl.java:839) ~[main/:?]
	at org.apache.ignite.internal.sql.engine.exec.ExecutionServiceImpl.onMessage(ExecutionServiceImpl.java:523) ~[main/:?]
	at org.apache.ignite.internal.sql.engine.exec.ExecutionServiceImpl.lambda$start$2(ExecutionServiceImpl.java:277) ~[main/:?]
	at org.apache.ignite.internal.sql.engine.message.MessageServiceImpl.onMessageInternal(MessageServiceImpl.java:151) ~[main/:?]
	at org.apache.ignite.internal.sql.engine.message.MessageServiceImpl.lambda$onMessage$1(MessageServiceImpl.java:117) ~[main/:?]
	at org.apache.ignite.internal.sql.engine.exec.QueryTaskExecutorImpl.lambda$execute$0(QueryTaskExecutorImpl.java:83) ~[main/:?]
	... 3 more
```
The test doesn't fail but, as I understand, there is a NPE-detector on TeamCity which may not like this.

Reply:
IntelliJ's static analyser does not like that line of code either. It expects an NPE here.
I guess it is time to fix it by adding some guard condition.